### DISCUSSION TITLE
v0.1.0-preview4

### DISCUSSION THREAD

**Discussion by jborean93** [jborean93/PSOpenAD#39] [Announcements]
## v0.1.0-preview4 - 2022-06-15

+ Added error handling for search request that ends with a referral
  + Currently the cmdlet will emit an error record with the referral URI which is similar to what the AD cmdlets do
+ Have exceptions in the background recv thread tasks bubble up as inner exceptions to preserve the stack trace for better debugging
+ Fix authentication with explicit credential on Windows
+ Added `-TracePath` to `New-OpenADSessionOption` to help debug raw LDAP traffice exchanged in a session.
+ Fix credential prompt when specifying `-Credential my-username` for a `PSCredential` parameter
+ Have `Get-OpenADWhoami` return an object with more details on the LDAP session, like the domain controller DNS name, URI, and authentication method used.
  + The returned username value will also strip the leading `u:` prefix if it is present
+ Added the `DomainController` property to the `OpenADSession` class to help identify the domain controller the session is connected to
+ Fixed the default parameter sets of the `Get-OpenAD*` cmdlets to always use the default LDAP filter that selects all of that type unless an explicit filter or identity was provided
+ Added `-ClientCertificate` to `New-OpenADSessionOption` that is used to authenticate using a client X.509 certificate
+ Raise `UnpackLDAPMessageException` when failing to unpack a response from the server.
  + The exception contains the `LDAPMessage` property which is the raw byte string that was being unpacked.
+ Added the `DomainController` property to the results of any `Get-OpenAD*` objects to help identify what domain controller returned that information

<hr /><em>This discussion was created from the release <a href='https://github.com/jborean93/PSOpenAD/releases/tag/v0.1.0-preview4'>v0.1.0-preview4</a>.</em>

**Comment by jdhitsolutions** [comment]
`Get-OpenADWhoAmI` is much more useful now.

**Reply by jborean93** [reply]
Yea, plus it allows me to add more properties in the future if need be without breaking the existing API. Thanks for the suggestion on that.
